LibWeb/HTML: Resolve image decoding promises inside tasks

AtkinsSJ opened the pull request:
This corresponds to https://github.com/whatwg/html/pull/10753

While I was here, I've also implemented the check for the current request changing, and reformatted the spec comments to match our style.

ADKaster commented on Userland/Libraries/LibWeb/HTML/HTMLImageElement.cpp:
```diff
@@ -311,55 +314,97 @@ WebIDL::ExceptionOr<JS::NonnullGCPtr<WebIDL::Promise>> HTMLImageElement::decode(
+                CurrentRequestCompletelyAvailable,
+            };
+            ImageState image_state;
+            auto previous_state = current_request().state();
```
This double event loop spin has always been truly puzzling. ... why do we need it?

AtkinsSJ replied:
I'm confused, I only see one spin here.

ADKaster commented on Userland/Libraries/LibWeb/HTML/HTMLImageElement.cpp:
```diff
@@ -311,55 +314,97 @@ WebIDL::ExceptionOr<JS::NonnullGCPtr<WebIDL::Promise>> HTMLImageElement::decode(
+                CurrentRequestBroken,
+                CurrentRequestCompletelyAvailable,
+            };
+            ImageState image_state;
```
:/ not initialized at declaration.

ADKaster commented on Userland/Libraries/LibWeb/HTML/HTMLImageElement.cpp:
```diff
@@ -311,55 +314,97 @@ WebIDL::ExceptionOr<JS::NonnullGCPtr<WebIDL::Promise>> HTMLImageElement::decode(
-                return !this->document().is_fully_active() || state == ImageRequest::State::Broken || state == ImageRequest::State::CompletelyAvailable;
-            }));
+                if (state != previous_state) {
+                    image_state = ImageState::CurrentRequestChanged;
```
does the state of the current request changing really mean that the current request changed? What about ABA? Should we be storing the "Current request" and making sure the image element still has the same request instead?

ADKaster replied:
"or is mutated" :thinking: How do we detect that?

AtkinsSJ commented:
Changes:

- Replaced the "spin and set an enum, then act on the enum" code with "just do the action in the spin callback".
- Removed the "request changed or was mutated" check for now, because it's clear I was not actually doing that correctly before, and it's not at all obvious what the correct thing is.

tcl3 commented:
This has conflicts.

AtkinsSJ commented:
Fixed the conflict and updated to use the new LibGC renamed methods. (I may have forgotten about this PR... 😅)
